tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -377,142 +377,116 @@ public:
-                                 IsDataContainer<X3>::value,
-                              int> = 0>
-   void Exec(unsigned int slot, const X0 &x0s, const X1 &x1s, const X2 &x2s, const X3 &x3s)
+   template <std::size_t ColIdx, typename End_t, typename... Its>
+   void ExecLoop(unsigned int slot, End_t end, Its... its)
    {
-      auto thisSlotH = fObjects[slot];
-      if (!(x0s.size() == x1s.size() && x1s.size() == x2s.size() && x1s.size() == x3s.size())) {
-         throw std::runtime_error("Cannot fill histogram with values in containers of different sizes.");
-      }
-      auto x0sIt = std::begin(x0s);
-      const auto x0sEnd = std::end(x0s);
-      auto x1sIt = std::begin(x1s);
-      auto x2sIt = std::begin(x2s);
-      auto x3sIt = std::begin(x3s);
-      for (; x0sIt != x0sEnd; x0sIt++, x1sIt++, x2sIt++, x3sIt++) {
-         thisSlotH->Fill(*x0sIt, *x1sIt, *x2sIt, *x3sIt); // TODO: Can be optimised in case T == vector<double>
+      auto *thisSlotH = fObjects[slot];
+      // loop increments all of the iterators while leaving scalars unmodified
+      // TODO this could be simplified with fold expressions or std::apply in C++17
+      auto nop = [](auto &&...) {};
+      for (; GetNthElement<ColIdx>(its...) != end; nop(Increment(its)...)) {
+         // dereference iterators, but pass scalars through unmodified
+         thisSlotH->Fill(Deref(its)...);
       }
    }
 
-   template <typename X0, typename X1, typename X2, typename W,
-             std::enable_if_t<IsDataContainer<X0>::value && IsDataContainer<X1>::value && IsDataContainer<X2>::value &&
-                                 !IsDataContainer<W>::value,
-                              int> = 0>
-   void Exec(unsigned int slot, const X0 &x0s, const X1 &x1s, const X2 &x2s, const W w)
+   // at least one container argument
+   template <typename... Xs,
+             typename std::enable_if<!Conjunction<std::is_convertible<Xs, double>...>::value, int>::type = 0>
+   void Exec(unsigned int slot, const Xs &...xs)
    {
-      auto thisSlotH = fObjects[slot];
-      if (!(x0s.size() == x1s.size() && x1s.size() == x2s.size())) {
-         throw std::runtime_error("Cannot fill histogram with values in containers of different sizes.");
-      }
-      auto x0sIt = std::begin(x0s);
-      const auto x0sEnd = std::end(x0s);
-      auto x1sIt = std::begin(x1s);
-      auto x2sIt = std::begin(x2s);
-      for (; x0sIt != x0sEnd; x0sIt++, x1sIt++, x2sIt++) {
-         thisSlotH->Fill(*x0sIt, *x1sIt, *x2sIt, w);
+      // array of bools keeping track of which inputs are containers
+      constexpr std::array<bool, sizeof...(Xs)> isContainer = {{!std::is_convertible<Xs, double>::value...}};
+
+      // index of the first container input
+      constexpr std::size_t colidx = FindIdxTrue(isContainer);
+      static_assert(colidx < sizeof...(Xs), "Error: index of collection-type argument not found.");
```

Comment:
I don't see how this can ever happen?

Reply:
Yah it shouldn't, but the static_assert doesn't really hurt here...

Reply:
it makes a reader stop and think "waaait, this can happen!? I must have missed something" :smile:

if you want to keep the check for robustness please add a comment along the lines of `// if this happens, there is a bug in the implementation`.